File: app/src/main/java/com/ps/simplepapersoccer/GameObjects/Game/GameHandler.java
package com.ps.simplepapersoccer.GameObjects.Game;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.UUID;

import com.ps.simplepapersoccer.AI.GameAIHandler;
import com.ps.simplepapersoccer.Activities.GameActivity;
import com.ps.simplepapersoccer.Enums.DifficultyEnum;
import com.ps.simplepapersoccer.Enums.NodeTypeEnum;
import com.ps.simplepapersoccer.Enums.VictoryConditionEnum;
import com.ps.simplepapersoccer.GameObjects.Move.PartialMove;
import com.ps.simplepapersoccer.GameObjects.Move.PossibleMove;
import com.ps.simplepapersoccer.GameObjects.Player;
import com.ps.simplepapersoccer.Helpers.MathHelper;
import com.ps.simplepapersoccer.R;
import com.google.common.collect.Iterables;

public class GameHandler {
	public Player player1;
	public Player player2;

	public Player currentPlayersTurn;

	public int numberOfTurns = 0;

	private GameActivity gameActivity;
	private GameAIHandler gameAIHandler;
	private GameBoard gameBoard;

	private boolean isMultiplayer;
	public boolean aiTurn = false;

	public Node ballNode() { return gameBoard.ballNode; }

	@Override
	public int hashCode() {
		return currentPlayersTurn.hashCode() ^ gameBoard.hashCode();
	}
	
	public GameHandler(final GameActivity gameActivity, int gridX, int gridY, DifficultyEnum difficulty, ArrayList<Player> players, boolean isMultiplayer)
	{
		this.player1 = players.get(0);
		this.player2 = players.get(1);

		currentPlayersTurn = players.get(0);

		this.isMultiplayer = isMultiplayer;

		gameBoard = new GameBoard(gridX, gridY);
		player1.goalNode = gameBoard.goalNode1;
		player2.goalNode = gameBoard.goalNode2;
		
		this.gameActivity = gameActivity;
		gameActivity.reDraw();

		gameAIHandler = new GameAIHandler(this, difficulty);
	}

	public HashSet<PossibleMove> allPossibleMovesFromNode(Node node)
	{
		return gameBoard.allPossibleMovesFromNode(node);
	}

	public void UpdateGameState()
	{
		gameActivity.UpdateDrawData();

		if(isGameOver())
		{
			winner(getWinner(ballNode()));
			return;
		}
		if (currentPlayersTurn.isAi && !isMultiplayer)
		{
			aiTurn = true;
			gameAIHandler.MakeAIMove();
		};
	}

	public void PlayerMakeMove(Node node, Player player)
	{
		PartialMove partialMove = new PartialMove(ballNode(), node, player);
		if (isPartialMoveLegal(partialMove, player) && !aiTurn)
		{
			MakeMove(partialMove);
			UpdateGameState();
		}
	}

	public void AIMakeMove(PartialMove move)
	{
		MakeMove(move);
		aiTurn = false;
		UpdateGameState();
	}

	public void MakeMove(PartialMove partialMove)
	{
		MakePartialMove(partialMove);

		if(ballNode().nodeType != NodeTypeEnum.Empty) {
			gameActivity.fxPlayer.playSound(R.raw.bounce);
		}
		else {
			gameActivity.fxPlayer.playSound(R.raw.soccerkick);
		}

		gameActivity.DrawPartialMove(partialMove, partialMove.madeTheMove.playerColor);
		++numberOfTurns;
	}

	public void UndoLastMove()
    {
		currentPlayersTurn = gameBoard.UndoLastMove().madeTheMove;
    }

	public void MakePartialMove(PartialMove partialMove)
	{
		gameBoard.MakePartialMove(partialMove);

		if (partialMove.newNode.nodeType == NodeTypeEnum.Empty) {
			changeTurn();
		}
	}

	public boolean isGameOver()
	{
		if (getWinner(gameBoard.ballNode) != null)
		{
			return true;
		};
		return false;
	}

	public Victory getWinner(Node node)
	{
		if(node.nodeType == NodeTypeEnum.Goal)
		{
			if (node.id == currentPlayersTurn.goalNode.id)
			{
				return new Victory(getOpponent(currentPlayersTurn), VictoryConditionEnum.Goal);
			}
			return new Victory(currentPlayersTurn, VictoryConditionEnum.Goal);
		}	
		else if (node.neighbors.size() == 0)
		{
			return new Victory(getOpponent(currentPlayersTurn), VictoryConditionEnum.OpponentOutOfMoves);
		}
		return null;
	}
	
	//Let the activity know we have a winner
	public void winner(Victory victory)
	{
		victory.winner.score += 1;
		gameActivity.Winner(victory);
	}
		
	//Converts screen coordinates to node coordinates
	public Node coordsToNode(float x, float y)
	{
		x = (x+gameActivity.gameView.gridXdraw/2)/gameActivity.gameView.gridXdraw * gameActivity.gameView.gridXdraw;
		y = (y+gameActivity.gameView.gridYdraw/2)/gameActivity.gameView.gridYdraw * gameActivity.gameView.gridYdraw-gameActivity.gameView.topEdge;
		
		x = x/gameActivity.gameView.gridXdraw;
		y = y/gameActivity.gameView.gridYdraw;

		return gameBoard.findNodeByXY((int)x, (int)y);
	}

	public Player getOpponent(Player myPlayer)
	{
		if (myPlayer == player1) return player2;
		return player1;
	}

	public Player changeTurn()
	{
		currentPlayersTurn = getOpponent(currentPlayersTurn);
		return currentPlayersTurn;
	}

	public boolean isPartialMoveLegal(PartialMove partialMove, Player player)
	{
		HashSet<PossibleMove> test = allPossibleMovesFromNode(ballNode());
		return test.contains(new PossibleMove(partialMove.oldNode, partialMove.newNode)) && player == currentPlayersTurn;
	}
}

File: app/src/main/java/com/ps/simplepapersoccer/GameObjects/Game/GameHandler.kt
package com.ps.simplepapersoccer.gameObjects.game

import android.os.Handler
import com.ps.simplepapersoccer.ai.GameAIHandler
import com.ps.simplepapersoccer.ai.abstraction.IGameAiHandlerListener
import com.ps.simplepapersoccer.data.enums.NodeTypeEnum
import com.ps.simplepapersoccer.data.enums.VictoryConditionEnum
import com.ps.simplepapersoccer.gameObjects.game.geometry.Node
import com.ps.simplepapersoccer.gameObjects.move.PartialMove
import com.ps.simplepapersoccer.gameObjects.player.AIPlayer
import com.ps.simplepapersoccer.gameObjects.player.abstraction.IPlayer
import kotlinx.coroutines.CoroutineDispatcher

class GameHandler(private val listener: IGameHandlerListener?, gridX: Int, gridY: Int,
                  private val players: ArrayList<IPlayer>,
                  private val dispatcher: CoroutineDispatcher, private val handler: Handler = Handler()): IGameAiHandlerListener {

    private val player1: IPlayer = players[0]
    private val player2: IPlayer = players[1]

    val currentPlayersTurn: IPlayer get() = players.first { it.playerNumber == gameBoard.currentPlayersTurn }

    var winner: Victory? = null

    var numberOfTurns = 0
    val gameBoard = GameBoard(gridX * 2, gridY * 2)
    private var ongoingTurn = false

    val ballNode: Node get() = gameBoard.ballNode

    override fun hashCode(): Int {
        return currentPlayersTurn.hashCode() xor gameBoard.hashCode()
    }

    init {
        gameBoard.currentPlayersTurn = players[0].playerNumber
        player1.goal = gameBoard.goal1
        player2.goal = gameBoard.goal2
        listener?.reDrawLiveData?.value = true
    }

    fun updateGameState() {
        ongoingTurn = false
        if (isGameOver) {
            winner(getWinner(ballNode)!!)
            return
        }
        if (currentPlayersTurn.isAi) {
            GameAIHandler(this, dispatcher, handler).makeAIMove(currentPlayersTurn as AIPlayer, this)
        }
    }

    override fun aiMove(partialMove: PartialMove?, timedOut: Boolean) {
        if (timedOut || partialMove == null || isPartialMoveLegal(partialMove, currentPlayersTurn).not()) {
            winner(Victory(getOpponent(currentPlayersTurn), if (partialMove == null) VictoryConditionEnum.IllegalMove else VictoryConditionEnum.TimedOut))
        } else {
            aiMakeMove(partialMove)
        }
    }

    fun playerMakeMove(node: Node, player: IPlayer) {
        val partialMove = PartialMove(ballNode, node, player.playerNumber)
        if (isPartialMoveLegal(partialMove, player) && currentPlayersTurn == player && !ongoingTurn) {
            ongoingTurn = true
            makeMove(partialMove)
        }
    }

    fun aiMakeMove(move: PartialMove) {
        ongoingTurn = true
        makeMove(move)
    }

    private fun makeMove(partialMove: PartialMove) {
        gameBoard.makePartialMove(partialMove)
        listener?.drawPartialMoveLiveData?.value = partialMove
        ++numberOfTurns
        updateGameState()
    }

    val isGameOver: Boolean
        get() {
            if (getWinner(gameBoard.ballNode) != null) {
                return true
            }
            return false
        }

    fun getWinner(node: Node): Victory? {
        if (node.nodeType == NodeTypeEnum.Goal) {
            if (player1.goal!!.isGoalNode(node)) return Victory(player2, VictoryConditionEnum.Goal)
            else if (player2.goal!!.isGoalNode(node)) return Victory(player1, VictoryConditionEnum.Goal)
        } else if (node.neighbors.size == 0) {
            return Victory(getOpponent(currentPlayersTurn), VictoryConditionEnum.OpponentOutOfMoves)
        }
        return null
    }

    private fun winner(victory: Victory) {
        this.winner = victory
        victory.winner.score += 1
        listener?.winnerLiveData?.value = victory
    }

    fun getOpponent(myPlayer: IPlayer): IPlayer {
        return when (myPlayer) {
            player1 -> player2
            player2 -> player1
            else -> player2
        }
    }

    fun getPlayerPosition(player: IPlayer): Int {
        return players.indexOf(player)
    }

    private fun isPartialMoveLegal(partialMove: PartialMove, player: IPlayer): Boolean {
        return gameBoard.allPossibleMovesFromNode(ballNode).any { x -> x.newNode == partialMove.newNode }
                && player == currentPlayersTurn
    }

    companion object {
        private val TAG = GameHandler::class.java.canonicalName
    }
}
